internal/pkg/agent/cmd/unprivileged.go

```diff
@@ -0,0 +1,127 @@
+unprivileged it will still perform all the same work, including stopping and starting the Elastic Agent.
+`,
+		Args: cobra.ExactArgs(0),
+		Run: func(c *cobra.Command, args []string) {
+			if err := unprivilegedCmd(streams, c); err != nil {
+				fmt.Fprintf(streams.Err, "Error: %v\n%s\n", err, troubleshootMessage())
+				os.Exit(1)
+			}
+		},
+	}
+
+	cmd.Flags().BoolP("force", "f", false, "Do not prompt for confirmation")
+	cmd.Flags().DurationP("daemon-timeout", "", 0, "Timeout waiting for Elastic Agent daemon")
+
+	return cmd
+}
+
+func unprivilegedCmd(streams *cli.IOStreams, cmd *cobra.Command) (err error) {
+	isAdmin, err := utils.HasRoot()
+	if err != nil {
+		return fmt.Errorf("unable to perform unprivileged command while checking for root/Administrator rights: %w", err)
+	}
+	if !isAdmin {
+		return fmt.Errorf("unable to perform unprivileged command, not executed with %s permissions", utils.PermissionUser)
+	}
+
+	// TODO(blakerouse): More work to get this working on macOS.
+	// Need to switch the vault from keystore based to file based vault.
+	if runtime.GOOS == "darwin" {
+		return errors.New("unable to perform unprivileged on macOS (not supported)")
+	}
+
+	// cannot switch to unprivileged when Elastic Defend exists in the policy
+	err = ensureNoElasticDefend()
+	if err != nil {
+		// error already adds context
+		return err
+	}
+
+	topPath := paths.Top()
+	daemonTimeout, _ := cmd.Flags().GetDuration("daemon-timeout")
+	force, _ := cmd.Flags().GetBool("force")
+	if !force {
+		confirm, err := cli.Confirm("This will restart the running Elastic Agent and convert it to run in unprivileged mode. Do you want to continue?", true)
+		if err != nil {
+			return fmt.Errorf("problem reading prompt response")
+		}
+		if !confirm {
+			return fmt.Errorf("unprivileged switch was cancelled by the user")
+		}
+	}
+
+	pt := install.CreateAndStartNewSpinner(streams.Out, "Converting Elastic Agent to unprivileged...")
+	err = install.SwitchExecutingMode(topPath, pt, install.ElasticUsername, install.ElasticGroupName)
+	if err != nil {
+		// error already adds context
+		return err
+	}
+
+	// wait for the service
+	if daemonTimeout >= 0 {
+		pt.Describe("Waiting for running service")
+		ctx := handleSignal(context.Background()) // allowed to be cancelled
+		err = wait.ForAgent(ctx, daemonTimeout)
+		if err != nil && !errors.Is(err, context.Canceled) {
+			pt.Describe("Failed waiting for running service")
+			return err
+		}
+		pt.Describe("Service is up and running")
+	}
+
+	return nil
+}
+
+func ensureNoElasticDefend() error {
+	ctx := context.Background()
+	l, err := newErrorLogger()
+	if err != nil {
+		return fmt.Errorf("failed to create error logger: %w", err)
+	}
+	comps, err := getComponentsFromPolicy(ctx, l, paths.ConfigFile(), 0)
+	if err != nil {
+		return fmt.Errorf("failed to create component model from policy: %w", err)
+	}
+	for _, comp := range comps {
+		if comp.InputSpec != nil && comp.InputSpec.InputType == endpoint {
```

Comment:
Is there a way to actually evaluate the runtime preventions and see if something has the root one?

That would be a bit more future proof. The constrain isn't really "is defend running" it's "is there something that requires root to function" running, which today is only defend.

Reply:
I think we need to determine if that is the behavior that we actually want. An input specification that requires root, when switched to unprivileged mode will still allow the Elastic Agent to run and the input will just report as failed. This would then be viewable to the user in `status` as well as in Fleet and would allow them to adjust/remove the integration.

In the case of Endpoint it is more detrimental to switch the mode with it enabled. Endpoint being a separate piece of software that is installed, once switched it cannot be uninstalled. That leaves the Elastic Agent in a completely broken state, unless switched back to privileged and then remove the integration.

I believe that we have 4 options for handling this:

1. Prevent switch if any input has a runtime protection for requiring root.
2. Prevent switch if any service input has a runtime protection for requiring root.
3. Add new option to input specification that says it doesn't support unprivileged mode directly. This is a little different then a runtime protection, being that it could be used in switch to prevent the switch. Endpoint specification can set this specifically and then future inputs can as well.
4. Keep it how it is at the moment and hard-code it to endpoint.

Reply:
> That would be a bit more future proof. The constrain isn't really "is defend running" it's "is there something that requires root to function" running, which today is only defend.

Osquery as well

Reply:
> 3. Add new option to input specification that says it doesn't support unprivileged mode directly.

Let's avoid building something new for information we already have, so this option is out.

> This would then be viewable to the user in status as well as in Fleet and would allow them to adjust/remove the integration.

Agree with this, along the lines of not building anything new, we can rely on this to inform the user after the fact. When this is exposed from Fleet, Fleet will have information about the installed packages and will likely warn or prevent the user in the same way. Most users are going to do this en-mass at won't be reading warning text anyway.

> Prevent switch if any service input has a runtime protection for requiring root.
> Keep it how it is at the moment and hard-code it to endpoint.

Given the distinction that a service input may be totally broken if we do this, it makes sense to only add additional functionality for service components to deal with this distinction.

If it isn't a lot of effort to check the spec file so we don't have to hard code this case, then doing this in a general way is preferable.

Otherwise given endpoint is the only service component we expect to have for the foreseeable future, we can leave this as is and add a comment explaining why it specifically is special cased.

Reply:
> > That would be a bit more future proof. The constrain isn't really "is defend running" it's "is there something that requires root to function" running, which today is only defend.
> 
> Osquery as well

@aleksmaus Root is required to function or it doesn't work well without root? Trying to understand the difference here.

Reply:
Another option is to allow the switch with Endpoint enabled, but Elastic Agent uninstall Endpoint just like the `uninstall` command does. It can perform uninstall and then switch to non-root. This provides more of a way of not blocking the user from switching, we could even prompt if they want to remove it an switch. [y/N] could be the default with being a force `-f` switch where it always uninstalls it.

I like this option a little more because it doesn't prevent the switch just because a service component that requires root is enabled.

Reply:
> @aleksmaus Root is required to function or it doesn't work well without root? Trying to understand the difference here.

We could run osqueryd binary standalone without root permissions and some of the functionality will not be available and some of the queries will return different results (with no errors). 

So there is no indication from the user point of view that something is wrong with permissions, the data just would be missing, which creates "blind spots" from security point of view without user being aware. 

The file system searches would be missing some valid results because of the lack of the permissions, and osquery doesn't return the error in these cases, it just returns a smaller resultset.

Here are just a few examples of running osquery with sudo and without on the same query:

<img width="1409" alt="Screenshot 2024-05-20 at 9 52 11 AM" src="https://github.com/elastic/elastic-agent/assets/872351/a71d757f-18be-456d-837f-01bb60906859">


<img width="1723" alt="Screenshot 2024-05-20 at 10 02 48 AM" src="https://github.com/elastic/elastic-agent/assets/872351/7c813abf-a25d-4acd-ab70-04f9b9302d80">


<img width="1511" alt="Screenshot 2024-05-20 at 10 04 02 AM" src="https://github.com/elastic/elastic-agent/assets/872351/ca9cb129-4d7f-4351-beba-4f7e15b666a3">


<img width="1403" alt="Screenshot 2024-05-20 at 10 13 00 AM" src="https://github.com/elastic/elastic-agent/assets/872351/84e2bd2d-1614-4e82-85b5-b7e98ac27f12">


Also the file permissions for our extensions should be set to root or Administrator:
https://osquery.readthedocs.io/en/stable/deployment/extensions/#autoloading-extensions
```
(extension executable files must be owned by root or Administrator)
```

Reply:
@aleksmaus I think it's best to move this to its own issue and discussion on osquerybeat as non-root. Its could be possible to allow it to run or not run. I believe that discussion is better to be had in a specific issue and then come to a conclusion. We can adjust in a follow up PR based on that.

Reply:
> Another option is to allow the switch with Endpoint enabled, but Elastic Agent uninstall Endpoint just like the uninstall command does. It can perform uninstall and then switch to non-root. This provides more of a way of not blocking the user from switching, we could even prompt if they want to remove it an switch. [y/N] could be the default with being a force -f switch where it always uninstalls it.

You would have to include and test that tamper protection still applies on this path which will be extra testing we have to keep around. I think it is much simpler if we force users to put unprivileged agents in policies that do not include Defend. There are no edge cases around Defend being conditionally uninstalled or removed that way. The composable agent policies work in the UI will help users deal with any shared content between the privileged and privileged policies.

So I'd say we should continue to refuse the conversion if defend is in the policy.

Reply:
Okay. I was just worrying that it prevents issues with different personas in an organization. A persona that manages the policies and a persona that manages installation of the endpoints. That is probably rare and not something we need to worry about.